File: 0087.clj
(require '(clojure.contrib [math :as math]))
(require '(libs [math :as my-math]))
(load-file "primes.clj")

; 11^2 + 2^3 + 2^4 = 145
; 2^2 + 5^3 + 2^4 = 145
; 210 

(def limit 50000000)
(def sqrtlimit (math/sqrt limit))
(println
 sqrtlimit
)
(def primes (sieve (math/floor sqrtlimit)))
(println
 primes
)

(println (time
(loop [c primes n #{}]
  (let [r (loop [b primes n n]
	    (let [r (loop [a primes n n]
		      (if (or (empty? a) (>= (+ (* (first a) (first a))
						(* (first b) (first b) (first b))
						(* (first c) (first c) (first c) (first c)))
					     limit))
			n
			(recur (rest a) (conj n
					      (+ (* (first a) (first a))
						 (* (first b) (first b) (first b))
						 (* (first c) (first c) (first c) (first c))
						 )))))]
	      (if (or (empty? (rest b)) (>= (+ (* (first primes) (first primes))
					       (* (first (rest b)) (first (rest b)) (first (rest b)))
					       (* (first c) (first c) (first c) (first c)))
					    limit))
		r
		(recur (rest b) r))))]
    (if (or (empty? (rest c)) (>= (+ (* (first primes) (first primes))
				     (* (first primes) (first primes) (first primes))
				     (* (first (rest c)) (first (rest c)) (first (rest c)) (first (rest c))))
				  limit))
      (count r)
      (recur (rest c) r))))
))

File: 0087.py
import primes
import math

if __name__ == '__main__':
    limit = 50000000
    sqrtlim = math.sqrt(limit)
    primes = primes.find_all_primes_less_than(sqrtlim)
    results = set()
    for c in primes:
        for b in primes:
            for a in primes:
                if a ** 2 + b ** 3 + c ** 4 < limit:
                    results.add(a ** 2 + b ** 3 + c ** 4)
                else:
                    break
            if c ** 4 + b ** 3 + primes[0] ** 2 >= limit:
                break
        if c ** 4 + primes[0] ** 3 + primes[0] ** 2 >= limit:
            break
    print len(results)
